Statistics

Ana06 opened the pull request:
Statistics on top of Events#index. Issue #973.

I added:

- Events state
- Scheduled (start_time set) vs. Unscheduled events of the confirmed events.
- Track Distribution of the confirmed events.

![image](https://cloud.githubusercontent.com/assets/16052290/15985032/798642c0-2fe0-11e6-9238-79401d544656.png)

Ana06 commented:
I've just added 	statistics on top of Registrations#index (also in issue #973):

- Number of registrations (including the new ones since you last logged in)
- Registration attended vs. Registration not attended
- Distribution of affiliation

![image](https://cloud.githubusercontent.com/assets/16052290/15987779/f646d77a-3035-11e6-8b1b-8b67be6a95c7.png)

For the affiliation, as it has no color, I created a function to generate different colours that are different specially for consecutive ones and that are always the same ones (to avoid confusion if they change every time the page is reloaded). I think it works really well:

![image](https://cloud.githubusercontent.com/assets/16052290/15987743/64c1accc-3034-11e6-8070-59551decaa48.png)

I show _None_, for the ones that have no affiliation, but maybe we want that the ones that do not have affiliation to not appear in the statistics to have more space for the rest.

Also, the registration, submissions and hours statistic in the dashboard are very similar. Maybe we could move it to a partial. I am using the first one so it would use the partial to avoid copying the code as I have done now. I am not sure if I should do it in this PR, as maybe it will make more difficult to review the rest.

I have a doubt. Why this two things don't do the same?

` reg.where.not(attended: true)`
 `reg.where(attended: false)`

The first one works as I expected, but the second doesn't and I don't understand why.

ChrisBr commented:
@Ana06 

> For the affiliation, as it has no color, I created a function to generate different colours that are different specially for consecutive ones and that are always the same ones (to avoid confusion if they change every time the page is reloaded). 

How do you make sure that it's always the same color?

ChrisBr commented:
> I show None, for the ones that have no affiliation, but maybe we want that the ones that do not have affiliation to not appear in the statistics to have more space for the rest.

I'd include it for now! We will see how useful it is in production and can adapt if necessary.

ChrisBr commented:
> I have a doubt. Why this two things don't do the same?

reg.where.not(attended: true)
reg.where(attended: false)

What I do in this case is to try it in the rails console (```rails c```). While executing the statements, rails will show you the generated and executed SQL statements. They're probably not the same as you get different results. Maybe this will already give you some insights to understand what's going wrong. If now, please paste the output of the rails console and we will try to debug with you.

ChrisBr commented:
> Registration attended vs. Registration not attended

This stastic makes only sense to display AFTER (maybe after the conference started) the conference...

ChrisBr commented on app/models/conference.rb:
```diff
@@ -317,6 +317,60 @@ def event_distribution
+  # * +hash+ -> hash
+  def scheduled_event_distribution
+    confirmed_events = program.events.where(state: 'confirmed')
+    scheduled_value =  { 'value' => confirmed_events.where('start_time IS NOT NULL').count, 'color' => 'green' }
```
I'd prefer to use here also ``where.not``

```
confirmed_events.where.not(start_time: nil)
```

hennevogel replied:
scope?

ChrisBr commented on app/models/conference.rb:
```diff
@@ -317,6 +317,60 @@ def event_distribution
+  def scheduled_event_distribution
+    confirmed_events = program.events.where(state: 'confirmed')
+    scheduled_value =  { 'value' => confirmed_events.where('start_time IS NOT NULL').count, 'color' => 'green' }
+    unscheduled_value =  { 'value' => confirmed_events.where('start_time IS NULL').count, 'color' => 'red' }
```
Same here!

ChrisBr commented on app/models/conference.rb:
```diff
@@ -317,6 +317,60 @@ def event_distribution
+    registrations.each do |r|
+      affiliations << (r.user.affiliation ? r.user.affiliation : 'None')
+    end
+    counted_affiliations = affiliations.group_by(&:capitalize).map {|k, v| [k, v.length]}
```
Hm, I'd prefer to do this in some ActiveRecord grouping? Because using database features is way faster than doing it in Ruby code.

Ana06 commented:
@ChrisBr 

> How do you make sure that it's always the same color?

Because they are not aleatory, that is why I needed a function that given  `i >= 1` gives you a different color for every `i` (**but always the same one for the same `i`**), and consecutive/close values of `i` gives highly different colors. The function name is `next_color`, I've used big prime numbers. II printed the first generated colors to show you that they are different and also because depending on the primes colors chosen the colours are more or less beautiful, I like that combination.

Ana06 commented:
@ChrisBr 

> What I do in this case is to try it in the rails console (rails c). While executing the statements, rails will show you the generated and executed SQL statements. They're probably not the same as you get different results. Maybe this will already give you some insights to understand what's going wrong. If now, please paste the output of the rails console and we will try to debug with you.

They are not the same, but I think they should be equivalent:

![image](https://cloud.githubusercontent.com/assets/16052290/16014440/64907a2c-3191-11e6-9d74-0339ef7c4360.png)

The only logical explanation would be that the attended value of the user that apear in the fist query but not in the second was `nil`, but it is not, it is `false`:

![image](https://cloud.githubusercontent.com/assets/16052290/16014577/0a05117a-3192-11e6-94b3-e8b31e6cb6c6.png)

Ana06 commented:
I'll change the _Registration attended vs. Registration not attended_ to be displayed only after the conference has started. I'll make the changes you added in the code too.

differentreality commented on app/views/admin/registrations/index.html.haml:
```diff
@@ -3,14 +3,34 @@
       %h1
         Registrations
-        = "(#{@registrations.length})" if @registrations
+        = "(#{@total_reg})" if @registrations
```
Would be awesome if that variable name wasn't abbreviated, so that I don't have to think twice when I am reading it.
`@total_reg`  also just holds `@registrations.count`, so do we really need to define a new variable for it?

Either way, since we have `@registrations` already loaded, we can use `.length` instead of `.count`, which triggers a new count query on our DB.

differentreality commented on app/models/conference.rb:
```diff
@@ -317,6 +317,57 @@ def event_distribution
+    result = {}
+    i=1
+    counted_affiliations.each do |key, value|
+      capitalized_key = (key.nil? ? 'None' : key.capitalize)
```
Affiliation, except for nil, might also just be an empty string. We still want to show 'None' for that. Perhaps check for `.blank?`

ChrisBr replied:
Isn't the nil case anyway covered by you query?

Ana06 replied:
The problem is that empty string  is grouped separated from nil. I can group them together

ChrisBr commented:
> Because they are not random, that is why I needed a function that given i >= 1 gives you a different color for every i (but always the same one for the same i), and consecutive/close values of i gives highly different colors. The function name is next_color, I've used big prime numbers. II printed the first generated colors to show you that they are different and also because depending on the primes colors chosen the colours are more or less beautiful, I like that combination.

> The colors could change if there are changes in the affiliation, but what I meant was that without having changes in the affiliation if you reload the page they'll be the sames (otherwise it would be confusing). They would have changed if they were random and also we may probably have different colors but maybe ugly ones.

-> That's what I meant. You can't be sure that the affiliation has always the same i. As long as you have thought about that issue it should be fine.

However, what we should maybe consider is grouping affiliations with less than a certain number together to other (e.g. less than 5 people are in that affiliation). Otherwise we will have A LOT affiliations only with one person which will make the pie diagram not very useful.

ChrisBr commented:
> I have a doubt. Why this two things don't do the same?
> reg.where.not(attended: true)
> reg.where(attended: false)

@Ana06 played a little bit around with that and it's indeed true that ``reg.where(attended: false)`` does not return a value. However, if you explicitely set the value to false

```
r = reg.where.not(attended: true).first
r.attended = false
r.save
```
and execute the query again, it works. Also if you select ``reg.where(attended: 0)`` if works. From the SQLite docu

> SQLite does not have a separate Boolean storage class. Instead, Boolean values are stored as integers 0 (false) and 1 (true)

So I assume, something goes wrong while setting the default value of attended.

ChrisBr commented on app/views/admin/registrations/index.html.haml:
```diff
@@ -3,14 +3,34 @@
+            = @new_reg
+  .col-md-4
+    = render partial: 'admin/conference/doughnut_chart', locals: {title: 'Affiliation', data: @affiliation_distribution}
+  = unless @conference.pending?
```
Shouldn't that be:
```
- unless @conference.pending?
```
?

Ana06 replied:
True, but it also works with `=`, because I tried it. Anyway, I'll change it.

Ana06 commented:
I've included the empty string in None, I've also group those affiliations that are less than the 2% of the registrations in others.

There is a problem, rubocop doesn't like that the Conference class has more that 500 lines.

differentreality commented:
Of course it does :)
Should we raise its limit again?

Ana06 commented:
@ChrisBr 

> However, if you explicitely set the value to false and execute the query again, it works. 

Are you sure that it works? Look:

![image](https://cloud.githubusercontent.com/assets/16052290/16038577/e5b8a16c-3225-11e6-9c6b-f2bc60805b64.png)

It's true that `reg.where(attended: 0)` works. Another reason why `reg.where(attended: false)` should work.

ChrisBr commented:
> Are you sure that it works? Look:
> 
> It's true that reg.where(attended: 0) works. Another reason why reg.where(attended: false) should work.

@Ana06 see this answer on stackoverflow: 

http://stackoverflow.com/questions/21066345/regression-in-activerecord-4-sqlite3-boolean-query-support

hennevogel commented:
@ChrisBr 
> This stastic makes only sense to display AFTER (maybe after the conference started) the conference...

I also want to know how I'm doing while the conference runs IMHO

hennevogel commented on app/models/conference.rb:
```diff
@@ -536,6 +597,16 @@ def registration_limit_exceeded?
   private
 
+  # Returns a different html colour for every i. We make use of big prime numbers
+  # to avoid repetition and to make consecutive colors clearly different.
```
You math studies are showing through ;-)

ChrisBr commented:
> >This stastic makes only sense to display AFTER (maybe after the conference started) the conference...

> I also want to know how I'm doing while the conference runs IMHO

That's why I wrote after the conference started!

hennevogel commented on app/views/admin/events/index.html.haml:
```diff
@@ -8,6 +8,13 @@
+    = render partial: 'admin/conference/doughnut_chart', locals: {title: 'Confirmed events scheduled', data: @scheduled_event_distribution}
+  .col-md-4
+    = render partial: 'admin/conference/doughnut_chart', locals: {title: 'Tracks of confirmed events', data: @tracks_distribution_confirmed}
+.row{ style: "margin-top: 40px;"}
```
Do not manipulate the box-model of bootstrap grid components.

Ana06 commented:
@hennevogel 

> I also want to know how I'm doing while the conference runs IMHO

Do you mean when the conference has started? It is shown if the start day of the conference is earlier or equal than today.

hennevogel commented on app/models/conference.rb:
```diff
@@ -317,6 +317,67 @@ def event_distribution
+  # * +hash+ -> hash
+  def scheduled_event_distribution
+    confirmed_events = program.events.where(state: 'confirmed')
+    scheduled_value =  { 'value' => confirmed_events.where.not(start_time: nil).count, 'color' => 'green' }
```
I don't like that it includes the color. This has clearly nothing to do with the model...

Ana06 replied:
and where do I put the color?

hennevogel replied:
Where you render this data. In the view :-)

Ana06 replied:
The rest doughnut charts set the color in the model too and doing it in the view would be complicated.

hennevogel replied:
OK

hennevogel commented on app/views/admin/registrations/index.html.haml:
```diff
@@ -11,6 +11,26 @@
         All the people who registered to your event
+.row
+  .col-sm-4.col-xs-4
+    .dashbox.text-center
```
Please do not mix `dashbox` and `doughnut_chart`. It looks horrible

Ana06 replied:
But I think it has no sense to show the registration with a `doughnut_chart`. Maybe we can remove the number of registrations from the statistics, as it is in the title.

Ana06 commented:
I eliminated the registrations dashbox, that is how it looks now:

![image](https://cloud.githubusercontent.com/assets/16052290/16196991/8a934d0c-3700-11e6-9acb-a3f21e98f0f4.png)

differentreality commented:
What is left to do here? Except of course for increasing the lines number for Conference class.

Would be very interesting to see the registration statistics any time now for oSC16.

Ana06 commented:
I think it is finished.

mdeniz commented on app/models/conference.rb:
```diff
@@ -536,6 +597,16 @@ def registration_limit_exceeded?
+    color[5..6] = ((i*151)%239 + 16).to_s(16)
+    color
+  end
+
```
Can be that way? For non math people :smile: 

```ruby
def next_color(i)
  "#" 
   + big_prime_color_component(:r, i) 
   + big_prime_color_component(:g, i) 
   + big_prime_color_component(:b, i)
end

def big_prime_color_component(component, i)
  big_prime_numbers = {r: 113, g: 67, b: 151}
  ((i*big_prime_numbers[component.to_sym])%239 + 16).to_s(16)
end
```

Ana06 commented:
Following @mdeniz  suggestion I made the next_color function more readable. Also I think I hadn't push the _Registrations dashbox eliminated from registrations_ commit, although I had added the screenshot. I've just added it.

hennevogel commented:
![giphy](https://cloud.githubusercontent.com/assets/514785/16361724/2e16454a-3b9a-11e6-8e3d-788229ba97a5.gif)

differentreality commented:
Let's add 10 extra lines for conference class so that rubocop doesn't complain.

Ana06 commented:
Rubocop doesn't complain now :relieved:

hennevogel commented:
Okay let's get this in and let's think about some other stats we can show in #973
